shell/server/permissions.js

```diff
@@ -0,0 +1,212 @@
+
+function PermissionSet(array) {
+  // A wrapper around an array of booleans.
+
+  if (!array) {
+    this.array = [];
+  } else if (array instanceof Array) {
+    this.array = array;
+  } else {
+    throw new Error("don't know how to interpret as PermissionSet: " + array);
+  }
+}
+
+// Methods for mutating a PermissionSet by combining it with another PermissionSet.
+// These return a boolean indicating whether the operation had any effect.
+
+PermissionSet.prototype.add = function(other) {
+  var changed = false;
+  for (var ii = 0; ii < other.array.length; ++ii) {
+    var old = this.array[ii];
+    this.array[ii] = this.array[ii] || other.array[ii];
+    if (old != this.array[ii]) {
+      changed = true;
+    }
+  }
+  return changed;
+}
+
+PermissionSet.prototype.remove = function(other) {
+  var changed = false;
+  for (var ii = 0; ii < other.array.length && ii < this.array.length; ++ii) {
+    var old = this.array[ii];
+    this.array[ii] = this.array[ii] && !other.array[ii];
+    if (old != this.array[ii]) {
+      changed = true;
+    }
+  }
+  return changed;
+}
+
+PermissionSet.prototype.intersect = function(other) {
+  var changed = false;
+  for (var ii = 0; ii < this.array.length; ++ii) {
+    var old = this.array[ii];
+    this.array[ii] = this.array[ii] && other.array[ii];
+    if (old != this.array[ii]) {
+      changed = true;
+    }
+  }
+  return changed;
+}
+
+function roleAssignmentPermissions(roleAssignment, viewInfo) {
+  var result = new PermissionSet([]);
+  if (!roleAssignment) {
+    return result;
+  }
+
+  if ("allAccess" in roleAssignment) {
+    var length = 0;
+    if (viewInfo.permissions) {
+      length = viewInfo.permissions.length;
+    }
+    var array = new Array(length);
+    for (var ii = 0; ii < array.length; ++ii) {
+      array[ii] = true;
+    }
+    result = new PermissionSet(array);
+  } else if ("roleId" in roleAssignment && viewInfo.roles && viewInfo.roles.length > 0) {
+    var roleDef = viewInfo.roles[roleAssignment.roleId];
+    if (roleDef) {
+      result = new PermissionSet(roleDef.permissions);
+    }
+  }
+
+  result.add(new PermissionSet(roleAssignment.addPermissionSet));
+  result.remove(new PermissionSet(roleAssignment.removePermissionSet));
+  return result;
+}
+
+function grainPermissionsInternal(grainId, userId, viewInfo) {
+  // Computes the permissions of the user by propagating permissions out from the grain's
+  // owner along role assignments until the graph is saturated.
+  //
+  // TODO(perf): This actually computes the permissions for *all* users. We should investigate ways
+  // to share results between calls. It may also make sense to cache some graph-connectivity
+  // information from the`mayOpenGrain()` computation to use here.
+
+  var grain = Grains.findOne(grainId);
+  var owner = grain.userId;
+
+  // userId -> PermissionSet
+  var permissionsMap = {}
+  permissionsMap[owner] = roleAssignmentPermissions({allAccess: null}, viewInfo);
+
+  var userStack = [owner];
```

Comment:
Hmm, I don't see why that would be. I'm pretty sure the algorithm is the same whether you're calculating owner -> recipient vs. recipient -> owner. In the owner -> recipient case your computing the set of permissions that flow to each node from the owner, with the ultimate answer being the set that flows to the recipient. In the recipient -> owner case you're computing the set of permissions that flow to the recipient _from_ each node, with the ultimate answer being the set of permissions that flow from the owner.

As a heuristic, if you keep track of the set of permissions that the recipient has already been proven to possess, then you can stop following edges that don't add any more permissions beyond that. In most cases, once you find any path from the owner to the recipient, you'll be able to quickly eliminate all other edges from consideration.

A further heuristic: Whenever inserting a new RoleAssignment, store its "distance" from the owner -- that is, the minimum number of edges that must be traversed from the owner to get to the user who is making the assignment. Then when you are later computing permissions, you can prefer to search RoleAssignments with lower distances first, to more quickly fill out the permissions set allowing you to eliminate other paths from consideration. (This optimization probably isn't necessary for now, though.)

There's probably a lot of other heuristics we could think of along these lines, too...

(FWIW, I spent several months working on a slightly different version of this problem at Google. Sadly, that code never made it into a product, because politics.)